closed/src/java.base/share/classes/openj9/internal/criu/CRIUSECProvider.java

```diff
@@ -39,26 +44,60 @@ public final class CRIUSECProvider extends Provider {
 
     private static final long serialVersionUID = -3240458633432287743L;
 
+    private static final class Action<T> extends SoftReference<T> {
+        private final Consumer<T> action;
+
+        Action(T object, Consumer<T> action) {
+            super(object);
+            this.action = action;
+        }
+
+        boolean reset() {
+            T object = get();
+
+            if (object != null) {
+                action.accept(object);
+                return true;
+            }
+
+            return false;
+        }
+    }
+
+    private static final List<Action<?>> links = new ArrayList<>();
+
+    public static <T> void doOnRestart(T object, Consumer<T> action) {
+        // associate a SoftReference to object with reset action
+        links.add(new Action<>(object, action));
+    }
```

Comment:
I think I gave you some bad advice. This may be better than what we had before, but it is missing the synchronization that would make it thread-safe and it still leaks. My new suggestion is to use a `WeakHashMap`:
```java
    private static final Map<Object, Consumer<Object>> actions = new WeakHashMap<>();

    @SuppressWarnings("unchecked")
    public static <T> void doOnRestart(T object, Consumer<T> action) {
        synchronized (actions) {
            // This unchecked cast is safe because the action
            // is only applied the supplied object.
            actions.put(object, (Consumer<Object>) action);
        }
    }

    public static void resetCRIUSEC() {
        synchronized (actions) {
            for (Map.Entry<Object, Consumer<Object>> entry : actions.entrySet()) {
                Object object = entry.getKey();

                if (object != null) {
                    entry.getValue().accept(object);
                }
            }
        }
    }
```

Reply:
Updated